### PR TITLE
Add a ZFS dashboard and alerts

### LINKED ISSUE #7761: Monitor ZFS Stats And Configure Alerts
### Problem

We need a way to monitor the usage of the zpool. We should investigate a solution to provide the following metrics(in order of importance):

1. Disk Capacity
2. IOPS
3. ZFS ARC statistics (size/usage/compression) - look into providing different stats for the two parts of the cache (MRU/MFU)

We should fire "soft" alerts for disk usage reaching 70%. Should fire a pager duty level alert if disk capacity ever reaches 75% on a pool. 

### Solution

The zfs localpv repo has information and links to a grafana dashboard that contains useful statistics. Investigate adding this dashboard (or a modified version) to our common helm chart. 
Potentially investigate other solutions to get metrics available on the nodes. Add zfs-localpv metrics via a PodMonitor or ServiceMonitor resource in common chart.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by jnels124**
**Description**:

* Add a flag to enable the Stackgres Prometheus exporter
* Add a new ZFS Grafana dashboard
* Add a `node` label to the node exporter metrics

**Related issue(s)**:

Fixes #7761

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: charts/hedera-mirror-importer/values.yaml [hunk 1]
```diff
@@ -229,6 +229,30 @@ prometheusRules:
       application: hedera-mirror-importer
       area: cloud
 
+  ImporterDbStorageFull:
```
Comment (jnels124) [inline charts/hedera-mirror-importer/values.yaml:232]: move to v2-values.yaml?

File: charts/hedera-mirror-importer/values.yaml [hunk 2]
```diff
@@ -229,6 +229,30 @@ prometheusRules:
...
+      summary: Database storage is full
+    enabled: true
+    expr: kubelet_volume_stats_used_bytes{node=~".*(worker|coord).*"} / kubelet_volume_stats_capacity_bytes{node=~".*(worker|coord).*"} >= .80
+    for: 1m
+    labels:
+      severity: critical
+      application: hedera-mirror-importer
+      area: resource
+
+  ImporterDbStorageFullWarning:
```
Comment (steven-sheehy) [inline charts/hedera-mirror-importer/values.yaml:244]: I don't think we need two alerts for the same thing. Can just drop this one.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 3]
```diff
@@ -0,0 +1,2084 @@
...
+      },
+      {
+        "current": {
+          "selected": false,
+          "text": "All",
+          "value": "$__all"
+        },
+        "datasource": {
+          "type": "prometheus",
+          "uid": "prometheus"
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:2040]: Don't hardcode the name as it differs on grafana cloud. Same for other variables.

```suggestion
          "type": "prometheus",
          "uid": "${prometheus}"
```

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 4]
```diff
@@ -0,0 +1,2084 @@
...
+        "options": [],
+        "query": {
+          "qryType": 1,
+          "query": "label_values(system_cpu_count,cluster)",
+          "refId": "PrometheusVariableQueryEditor-VariableQuery"
+        },
+        "refresh": 1,
+        "regex": "",
+        "skipUrlSync": false,
+        "sort": 0,
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:2057]: Sort the values ascending. Same for other variables.
```suggestion
        "sort": 1,
```

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 5]
```diff
@@ -0,0 +1,2084 @@
...
+  "refresh": "1m",
+  "schemaVersion": 39,
+  "tags": [],
+  "templating": {
+    "list": [
+      {
+        "current": {
+          "selected": false,
+          "text": "Loki",
+          "value": "P8E80F9AEF21F6940"
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:1922]: Don't hardcode the name.

```suggestion
          "text": "",
          "value": ""
```

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 6]
```diff
@@ -0,0 +1,2084 @@
...
+        "current": {
+          "selected": false,
+          "text": "All",
+          "value": "$__all"
+        },
+        "datasource": {
+          "type": "prometheus",
+          "uid": "${prometheus}"
+        },
+        "definition": "label_values(kubelet_volume_stats_used_bytes{node=~\"$node\"},namespace)",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:1968]:
```suggestion
        "definition": "label_values(kubelet_volume_stats_used_bytes{cluster=~\"$cluster\",node=~\"$node\"},namespace)",
```

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 7]
```diff
@@ -0,0 +1,2084 @@
...
+        "description": "List of nodes with ZFS pools",
+        "hide": 0,
+        "includeAll": true,
+        "label": "Node",
+        "multi": true,
+        "name": "node",
+        "options": [],
+        "query": {
+          "qryType": 1,
+          "query": "label_values(node_zfs_arc_size,node)",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:2013]: Filter by cluster and namespace

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 8]
```diff
@@ -0,0 +1,2084 @@
...
+            "x": 0,
+            "y": 2
+          },
+          "id": 54,
+          "links": [
+            {
+              "targetBlank": true,
+              "title": "Explore Logs",
+              "url": "/explore?panes={\"rl3\":{\"datasource\":\"${loki}\",\"queries\":[{\"refId\":\"A\",\"expr\":\"{app=\\\"importer\\\", cluster=\\\"${cluster:text}\\\", namespace=\\\"${namespace:text}\\\"} |= ``\",\"queryType\":\"range\",\"datasource\":{\"type\":\"loki\",\"uid\":\"${loki}\"},\"editorMode\":\"builder\"}],\"range\":{\"from\":\"${__from}\",\"to\":\"${__to}\"}}}&schemaVersion=1&orgId=1"
+            }
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:1873]: We can probably remove the custom link. Grafana has added built-in links to explore now in the hamburger menu on the right.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 9]
```diff
@@ -0,0 +1,2084 @@
...
+          ],
+          "options": {
+            "dedupStrategy": "none",
+            "enableLogDetails": false,
+            "prettifyLogMessage": false,
+            "showCommonLabels": true,
+            "showLabels": true,
+            "showTime": false,
+            "sortOrder": "Ascending",
+            "wrapLogMessage": false
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:1883]: The log panel is collapsed by default. I think we can expand it by default.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 10]
```diff
@@ -0,0 +1,2084 @@
...
+        "x": 0,
+        "y": 1
+      },
+      "id": 354,
+      "options": {
+        "legend": {
+          "calcs": [
+            "min",
+            "mean",
+            "max"
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:114]: The standard order in our other dashboards is `min, max, mean`. We should also sort mean by descending. Same comment for all graphs.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 11]
```diff
@@ -0,0 +1,2084 @@
...
+      "options": {
+        "legend": {
+          "calcs": [],
+          "displayMode": "list",
+          "placement": "bottom",
+          "showLegend": true
+        },
+        "tooltip": {
+          "mode": "single",
+          "sort": "none"
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:331]: Change to multi and desc. Same comment for others.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 12]
```diff
@@ -0,0 +1,2084 @@
...
+          "custom": {
+            "axisBorderShow": false,
+            "axisCenteredZero": false,
+            "axisColorMode": "text",
+            "axisLabel": "",
+            "axisPlacement": "auto",
+            "barAlignment": 0,
+            "drawStyle": "line",
+            "fillOpacity": 0,
+            "gradientMode": "none",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:274]: Same comment for others.

```suggestion
            "gradientMode": "hue",
```

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 13]
```diff
@@ -0,0 +1,2084 @@
...
+      "datasource": {
+        "type": "prometheus",
+        "uid": "prometheus"
+      },
+      "fieldConfig": {
+        "defaults": {
+          "color": {
+            "mode": "palette-classic"
+          },
+          "custom": {
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:265]: Disable grid lines on all graphs per our standard look and feel.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 14]
```diff
@@ -0,0 +1,2084 @@
...
+            "axisBorderShow": false,
+            "axisCenteredZero": false,
+            "axisColorMode": "text",
+            "axisLabel": "",
+            "axisPlacement": "auto",
+            "barAlignment": 0,
+            "drawStyle": "line",
+            "fillOpacity": 0,
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:273]: Same comment for others.

```suggestion
            "fillOpacity": 10,
```

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 15]
```diff
@@ -0,0 +1,2084 @@
...
+          "editorMode": "code",
+          "expr": "irate(node_disk_written_bytes_total{cluster=~\"$cluster\", node=~\"$node\", device=\"$zfsDeviceId\"}[$__rate_interval])",
+          "hide": false,
+          "instant": false,
+          "legendFormat": "write/{{node}}",
+          "range": true,
+          "refId": "B"
+        }
+      ],
+      "title": "Throughput",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:782]: Think this one and iops deserve a legend with min, max, mean. Move both to be full width.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 16]
```diff
@@ -0,0 +1,2084 @@
...
+          "editorMode": "code",
+          "expr": "irate(node_disk_writes_completed_total{cluster=~\"$cluster\", node=~\"$node\", device=\"$zfsDeviceId\"}[$__rate_interval])",
+          "hide": false,
+          "instant": false,
+          "legendFormat": "write/{{node}}",
+          "range": true,
+          "refId": "B"
+        }
+      ],
+      "title": "IOPS",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:890]: Add a description to all graphs.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 17]
```diff
@@ -0,0 +1,2084 @@
...
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "prometheus"
+          },
+          "editorMode": "code",
+          "expr": "kubelet_volume_stats_used_bytes{cluster=~\"$cluster\", node=~\"$node\", namespace=~\"$namespace\"}",
+          "instant": false,
+          "legendFormat": "{{node}}/{{namespace}}/{{persistentvolumeclaim}}",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:343]: A PVC ID is globally unique so we can shorten this to a reasonable length.
```suggestion
          "legendFormat": "{{persistentvolumeclaim}}",
```

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 18]
```diff
@@ -0,0 +1,2084 @@
...
+          "range": true,
+          "refId": "A"
+        },
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "${prometheus}"
+          },
+          "editorMode": "code",
+          "expr": "irate(node_disk_write_time_seconds_total{cluster=~\"$cluster\", node=~\"$node\", device=\"$zfsDeviceId\"}[$__rate_interval]) / irate(node_disk_writes_completed_total{node=~\"$node\", device=\"$zfsDeviceId\"}[$__rate_interval])",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:566]: We should filter out rows with missing data by adding either >0 or >=0. Same comment for other graphs.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: charts/hedera-mirror/values.yaml [hunk 19]
```diff
@@ -289,6 +289,7 @@ stackgres:
       wal_init_zero: "off"  # Not needed with ZFS
       wal_recycle: "off"  # Not needed with ZFS
       work_mem: "32MB"
+    enableMetricsExporter: false
```
[the remark this answers is not in the document]
Comment (jnels124) [reply charts/hedera-mirror/values.yaml:292]: keeping disabled until we resolve resource allocation

File: charts/hedera-mirror-common/dashboards/zfs.json (hunk 7 above)
Comment (jnels124) [reply charts/hedera-mirror-common/dashboards/zfs.json:2013]: filtering just on cluster but namespace is not possible. The namespace on this metric is going to be common because it is a shared resource.

File: charts/hedera-mirror-common/Chart.yaml [hunk 20]
```diff
@@ -31,7 +31,7 @@ dependencies:
     condition: zfs.enabled
     name: zfs-localpv
     repository: https://openebs.github.io/zfs-localpv
-    version: 2.4.1
+    version: 2.5.0
```
Comment (jnels124) [inline charts/hedera-mirror-common/Chart.yaml:34]: bumping zfs version here as very small task

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 21]
```diff
@@ -431,27 +441,29 @@
...
       "targets": [
         {
           "datasource": {
             "type": "prometheus",
-            "uid": "prometheus"
+            "uid": "${prometheus}"
           },
           "editorMode": "code",
-          "expr": "irate(node_disk_io_now{cluster=~\"$cluster\", node=~\"$node\", device=\"$zfsDeviceId\"}[$__rate_interval])",
+          "expr": "irate(node_disk_io_time_weighted_seconds_total{cluster=~\"$cluster\", node=~\"$node\", device=\"$zfsDeviceId\"}[$__rate_interval]) >= 0",
```
Comment (jnels124) [inline charts/hedera-mirror-common/dashboards/zfs.json:466]: this provides a more accurate representation of tasks that waited between scrape intervals

File: charts/hedera-mirror-importer/values.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply charts/hedera-mirror-importer/values.yaml]: This is still in importer chart

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 22]
```diff
@@ -0,0 +1,2102 @@
...
+      "targets": [
+        {
+          "datasource": {
+            "type": "prometheus",
+            "uid": "${prometheus}"
+          },
+          "editorMode": "code",
+          "expr": "irate(node_disk_io_time_weighted_seconds_total{cluster=~\"$cluster\", node=~\"$node\", device=\"$zfsDeviceId\"}[$__rate_interval]) >= 0",
+          "instant": false,
+          "legendFormat": "{{node}}",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:468]: This metric doesn't seem to have a node label in preprod so it's showing all labels in legend. Same comment for other `node_*` metrics. Is this just because it needs the prometheus relabeling change below first?

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: charts/hedera-mirror-common/dashboards/zfs.json (hunk 22 above)
Comment (jnels124) [reply charts/hedera-mirror-common/dashboards/zfs.json:468]: yes it needs the relabeling change first

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 23]
```diff
@@ -0,0 +1,2102 @@
...
+        "sortOrder": "Ascending",
+        "wrapLogMessage": false
+      },
+      "targets": [
+        {
+          "datasource": {
+            "uid": "${loki}"
+          },
+          "editorMode": "builder",
+          "expr": "{component=~\".*zfs.*\"}",
```
[the remark this answers is not in the document]
Comment (jnels124) [reply charts/hedera-mirror-common/dashboards/zfs.json:1922]: will include cluster but this is a common resource so will exclude namespace filter

[the remark this answers is not in the document]
Comment (jnels124) [reply charts/hedera-mirror-common/dashboards/zfs.json]: ended up not adding a rate panel to logs as there isn't a metric available to capture count of logged events for apps outside of spring

[the remark this answers is not in the document]
Comment (steven-sheehy) [reply charts/hedera-mirror-common/dashboards/zfs.json]: You can do it with pure logql, but no need to spend more time on it.

File: charts/hedera-mirror-common/dashboards/zfs.json [hunk 24]
```diff
@@ -2040,40 +2041,41 @@
...
-        "label": "Cluster",
+        "label": "Node",
         "multi": true,
-        "name": "cluster",
+        "name": "node",
         "options": [],
         "query": {
           "qryType": 1,
-          "query": "label_values(system_cpu_count,cluster)",
+          "query": "label_values(kube_pod_info{cluster=~\"$cluster\", pod=~\"$zfsComponentsRegex\"},node)",
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/dashboards/zfs.json:2068]: For Grafana Cloud, you have to pick metrics that are exported there. kube_pod_info is not. Is there another metric you can use? Double check the others

File: charts/hedera-mirror-common/values.yaml [hunk 25]
```diff
@@ -331,6 +331,20 @@ prometheus:
         cpu: 50m
         memory: 50Mi
 
+prometheusRules:
```
Comment (steven-sheehy) [inline charts/hedera-mirror-common/values.yaml:334]: Maybe move under `stackgres.prometheusRules` since they're not enabled unless stackgres is

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: charts/hedera-mirror-common/dashboards/zfs.json (hunk 24 above)
Comment (jnels124) [reply charts/hedera-mirror-common/dashboards/zfs.json:2068]: This dashboard is largely non functional in grafana cloud right now under current config. The only thing we send is things coming from job with values `.*(graphql|grpc|importer|monitor|rest|restjava|rosetta|traefik|web3)$`. We don't send any k8s related metrics for nodes or kube state. We can update the remote write to include the kube_pod_info from the kubestatemetrics job as well as zfs metrics if we decide we want them in grafana cloud

Comment (steven-sheehy) [reply charts/hedera-mirror-common/dashboards/zfs.json:2068]: While that's true, we shouldn't intentionally pick unrelated metrics that aren't needed for the graphs to populate the variables. We don't need to export the potentially large number of `kube_pod_info` time series just so we can populate the list of nodes.

Instead, I think you should use `kubelet_volume_stats_used_bytes` to populate all three cluster, namespace, and node variables since it's already used elsewhere within the dashboard.

Comment (jnels124) [reply charts/hedera-mirror-common/dashboards/zfs.json:2068]: I didn't intentionally pick an unrelated metric. There are going to be no other metrics that are going to provide a clear indication of zfs related pods unless we have a metric that has pod names and we don't have that outside of the mirror components. 

`kubelet_volume_stats_used_bytes `  doesn't indicate any information about what is running all the zfs components and neither does any other metric. The only definitive way to get that information is from pod names and we don't have those available for these components.

We can change back to the zfs  node metric it was using before to populate the node dropdown which will include only the zfs zpool nodes, then change the logs row to have a separate panel for the zfs controller that doesn't apply any of the dropdown filters.

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by xin-hedera — APPROVED** [review]
LGTM
